Add tensorboard support (pytorch, tf2+)

raubitsj opened the pull request:
This is ready to go when regressions pass: :)
 ./cling-regression.sh --cli_branch feature/tensorboard --spec wandb-examples-tensorboard

TODO: (in a follow up PR)
- [ ] Detect tf1 situations that we cant handle (right now we patch tensorboard but it doesnt get called)
- [ ] Try on all tf, pyt, and some tensorboardX situations?  make sure we have telemetry for failed environments (sentry?)
- [x] Add proto message for watching tb dirs
- [x] Implement the watcher
- [x] Save tensorboard files
- [x] Add cleanup routines
- [x] Add event consumer
- [x] Implement the summary -> history saving logic
- [ ] Figure out what needs to be done with reinit runs?  reset?
- [ ] Might be able to support tf1.15+ with a bit more work
- [ ] Move step calculation into internal process
- [ ] Support plugin images
- [ ] Add synchronous summary query interface in user process
- [ ] Add checks for local vs remote (when we have a chance to test remote files)
- [ ] Add unittests for watcher

Regression is basically passing: (except for things listed above that are not correct and test issues)
https://app.wandb.ai/jeffr/regression/groups/20200804-0.0.39-feature_tensorboard-a849f62-v257oz

vanpelt commented on wandb/framework/tensorboard/monkeypatch.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def _patch_nontensorflow(writer, module, save=None):
+    # This configures non-TensorFlow Tensorboard logging
```
non-TensorFlow or TensorFlow < 1.4x I believe.

vanpelt replied:
Or better, PyTorch, TensorboardX, or TensorFlow < 1.4x logging.

vanpelt replied:
I take both of these back as we're not actually patching old tensorflow with this, which I think is fine for initial release.

raubitsj replied:
yep.  i think we can make old tensorflow work with the watcher method, but there are more tricks needed like running event compat conversions.  Would like to defer a bit on this

vanpelt commented on wandb/framework/tensorflow/estimator_hook.py:
```diff
@@ -43,6 +42,6 @@ def after_run(self, run_context, run_values):
         print(run_values)
```
We actually have a print statement here?  Let's remove.

raubitsj replied:
removed

vanpelt commented on wandb/internal/tb_watcher.py:
```diff
@@ -0,0 +1,279 @@
+    elif not os.path.exists(wandb_path):
+        os.symlink(abs_path, wandb_path)
+    # TODO(jhr): need to figure out policy, live/throttled?
+    sender._save_file(file_name)
```
Looks familiar ;)  We should set `policy="live"` here.  It will upload every N seconds based on it's size.

raubitsj replied:
Added it, i was worried because cli-og has a throttled policy.  I left the note to look into policy improvements

vanpelt commented on wandb/internal/tb_watcher.py:
```diff
@@ -0,0 +1,279 @@
+class TBDirWatcher(object):
+    from tensorboard.backend.event_processing import directory_watcher  # type: ignore
+    from tensorboard.backend.event_processing import event_file_loader  # type: ignore
+    from tensorboard.compat import tf as tf_compat  # type: ignore
```
I'm a little worried about us not having tensorboard and how we handle the internal process crashing.  We should probably make an internal exception class that the process loop catches and returns an error object to the user process.

Also, wild idea, instead of doing the insane PTY console logic, we could have the user process queue up log messages from the backend process and flush them when any method on the run object is called...

raubitsj replied:
good catch, i thought imports inside classes wouldnt get called unless instantiated but i was confused.
converted to get_module.  

I agree with the internal process robustness need, I will look into that in a later PR.  At this point we shouldnt ever blow up in internal because we won't initialize this object unless the patch succeeded

vanpelt commented on wandb/internal/tb_watcher.py:
```diff
@@ -0,0 +1,279 @@
+    def __init__(self, tbwatcher, logdir, save, namespace, queue):
+        self._tbwatcher = tbwatcher
+        self._generator = self.directory_watcher.DirectoryWatcher(
+            logdir, self._loader(save, namespace), self._is_new_tensorflow_events_file
```
Something I haven't tried, but I think might *just work...  If `logdir` is a cloud storage url, the DirectoryWatcher uses the io.gfile implementation so I think it will poll cloud storage for changes.  We would need to make `_is_new_tensorflow_events_file` use [tf.io.exists(...)](https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/tensorflow_stub/io/gfile.py).  We would also want to try to pull in real tensorflow in place of tf_compat because the stock tensorboard implementation only supports S3.

The boss move would be to make a `wandb.io` library that mirrored [tf.io](https://github.com/tensorflow/tensorflow/tree/v2.3.0/tensorflow/python/lib/io) and we start compiling wheels.  Would make artifact references much simpler and enable some cool usecases.

raubitsj replied:
Wanted to wait on that until I had more time test non local files.

vanpelt commented on wandb/framework/tensorboard/monkeypatch.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def _notify_tensorboard_logdir(logdir, save=None):
+    wandb.run._tensorboard_callback(logdir, save=save)
```
For now if `logdir` is not a filesystem directory we should warn the user.  See later comments about cloud storage urls.

raubitsj replied:
I would like to defer this check until I can test non local files unless there is a safe check to do here.

vanpelt replied:
I think the current safe check is `os.path.exists(logdir)`

raubitsj replied:
I dont think I can count on the directory existing at this point

vanpelt replied:
The next best check would be "://" in logdir.

raubitsj replied:
ugly, but added :)

vanpelt commented on wandb/framework/tensorboard/log.py:
```diff
@@ -220,7 +220,8 @@ def log(tf_summary_str_or_pb, history=None, step=0, namespace="", **kwargs):
             RATE_LIMIT_SECONDS is None
-            or timestamp - STEPS["global"]["last_log"] >= RATE_LIMIT_SECONDS  # noqa: W503 E501
+            or timestamp - STEPS["global"]["last_log"]
+            >= RATE_LIMIT_SECONDS  # noqa: W503 E501
```
Also note we need to call `reset_state()` during `run.join()` cleanup to make STEPS reset.

raubitsj replied:
fixed nit,

i had a note that I was going to do this after redoing how we track _step, but I did it now for good measure, added reset_state() to TBWatcher init in internal.  I dont think the step logic works correctly as is. but reset wont make anything worse

vanpelt commented on wandb/framework/tensorboard/monkeypatch.py:
```diff
@@ -0,0 +1,64 @@
+def patch(save=None, tensorboardX=None, pytorch=None):
+    if len(wandb.patched["tensorboard"]) > 0:
+        raise ValueError(
+            "Tensorboard already patched, remove sync_tensorboard=True from wandb.init or only call wandb.tensorboard.patch once."
```
I take this back, we only explicitly call patch if we haven't already.  If the user is calling patch multiple times we should bomb out.

raubitsj replied:
yep, this is the same logic as cli-og, i think it is right

raubitsj commented on wandb/framework/tensorboard/monkeypatch.py:
```diff
@@ -0,0 +1,64 @@
+            writer=tb_writer, module=TENSORBOARD_PYTORCH_MODULE, save=save
+        )
+    else:
+        raise ValueError("Unsupported tensorboard configuration")
```
I would like to defer this work until we go through the full compatibilty matrix, i was focusing on latest pytorch and tensorflow right now, Check the TODO in the PR.

raubitsj commented:
@vanpelt , i addressed most of the comments, the main ones that were not addressed are best deferred until some more changes are made in a future PR.   The largest being better unittests,  I think there are many issues with step logging which need to be fixed before we could write unittests that will verify correct operation.
